### ISSUE TITLE
Does provider "async = true" need to be remove from url when register to registry?

### ISSUE THREAD

**Issue by fingthinking** [apache/incubator-dubbo#2321]
- [ ] I have searched the [issues](https://github.com/apache/incubator-dubbo/issues) of this repository and believe that this is not a duplicate.
- [ ] I have checked the [FAQ](https://github.com/apache/incubator-dubbo/blob/master/FAQ.md) of this repository and believe that this is not a duplicate.

### Environment

* Dubbo version:  2.7.0-SNAPSHOT

### Step to reproduce this issue

1. https://github.com/apache/incubator-dubbo/blob/master/dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/integration/RegistryProtocol.java
```java
 private URL getRegisteredProviderUrl(final Invoker<?> originInvoker) {
        URL providerUrl = getProviderUrl(originInvoker);
        //The address you see at the registry
        return providerUrl.removeParameters(getFilteredKeys(providerUrl))
                .removeParameter(Constants.MONITOR_KEY)
                .removeParameter(Constants.BIND_IP_KEY)
                .removeParameter(Constants.BIND_PORT_KEY)
                .removeParameter(QOS_ENABLE)
                .removeParameter(QOS_PORT)
                .removeParameter(ACCEPT_FOREIGN_IP)
                .removeParameter(VALIDATION_KEY)
                .removeParameter(INTERFACES);
    }
TO
 private URL getRegisteredProviderUrl(final Invoker<?> originInvoker) {
        URL providerUrl = getProviderUrl(originInvoker);
       // 此处移除async标记，仅保留标记在provider端
        Set<String> keySet = providerUrl.getParameters().keySet();
        List<String> asyncKey = new ArrayList<>();
        for(String key : keySet) {
            if(key.endsWith("." + Constants.ASYNC_KEY) || key.equals(Constants.ASYNC_KEY)) {
                asyncKey.add(key);
            }
        }
        //The address you see at the registry
        return providerUrl.removeParameters(getFilteredKeys(providerUrl))
                .removeParameter(Constants.MONITOR_KEY)
                .removeParameter(Constants.BIND_IP_KEY)
                .removeParameter(Constants.BIND_PORT_KEY)
                .removeParameter(QOS_ENABLE)
                .removeParameter(QOS_PORT)
                .removeParameter(ACCEPT_FOREIGN_IP)
                .removeParameter(VALIDATION_KEY)
                .removeParameter(INTERFACES)
                .removeParameters(asyncKey);
    }
```

that if Provider use "provider async" like this：
```Java
 public String sayHello(String name) {
        System.out.println("Main sayHello() method start.");
        final AsyncContext asyncContext = RpcContext.startAsync();
        new Thread(() -> {
            asyncContext.signalContextSwitch();
            System.out.println("Attachment from consumer: " + RpcContext.getContext().getAttachment("consumer-key1"));
            System.out.println("    -- Async start.");
            try {
                Thread.sleep(500);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            asyncContext.write("Hello " + name + ", response from provider.");
            System.out.println("    -- Async end.");
        }).start();
        System.out.println("Main sayHello() method end.");
        return "hello, " + name;
    }
 <dubbo:method name="sayHello" async="true"/>
```
and consumer user dubbo version low than 2.7.0-SNAPSHOT, 
"async = true" will know from provider registed url, that need consumer use old async to invoke this method. But it is not what we expected, we think async should only be set by consumer itself.
英文不太好，中文版：
1. Provider版本使用 2.7.0-SNAPSHOT版本，然后发布接口使用了"provider async"的方式（见上面第二段代码）,则需要在发布的时候添加` <dubbo:method name="sayHello" async="true"/>`则会将 async=true 的标记注册到注册中心，这时候就会将标记透传到 Consumer 端，导致服务调用变成异步的，而出现返回值为 null 的情况。
2. 根据我个人的理解，async 标记应该仅由 provider 或 consumer 自己指定，而不应该透传到对方感知。

**Comment by carryxyh** [comment]
Hi, I have check this case.

When using <dubbo:service async="true" .../>, the Consumer will not be affected. However, the Consumer call will be affected when using <dubbo:method name="sayHello" async="true"/>.

But I think the better solution is that Consumer ignores the async attribute of the Provider, instead of forbidding the provider to register the async attribute on the registry.

**Comment by diecui1202** [comment]
@carryxyh Do there have other scenarios to use the async attribute on consumer side?  If no, we can ignore it on consumer side.

**Comment by carryxyh** [comment]
@diecui1202 
No. On the consumer side should completely ignore the Provider's async attribute, so that the Consumer's async and Provider's async are completely isolated, so I ignored it in ClusterUtil in my pr.

**Comment by diecui1202** [comment]
@carryxyh I think we need to consider about the upgrading compatibility. Users need to upgrade the apps one by one, we can't ask them to do it with a particularly order. Even some times the Apps are both the consumer and provider.

**Comment by carryxyh** [comment]
Hi, @diecui1202 
I checked the use of method.async in the 2.6.x version, and method.async was not affected by the Provider in the old version. Suppose we update and restart the provider first. Since the consumer uses the old version, method.async does not affect the use of the Consumer. So we don't need to worry about compatibility issues

**Comment by fingthinking** [comment]
@carryxyh @diecui1202 https://github.com/apache/incubator-dubbo/blob/2.6.3-release/dubbo-rpc/dubbo-rpc-dubbo/src/main/java/com/alibaba/dubbo/rpc/protocol/dubbo/DubboInvoker.java

**Comment by carryxyh** [comment]
Hi, @diecui1202 
After looking at the code, I think we really have to consider compatibility issues. 

After thinking, I think it would be better if you fix this problem at both ends.
1. On the Provider side, the method.async is not exposed to ensure compatibility.
2. Ignore the Provider's method.async property on the Consumer side to ensure that the Provider's method.async does not affect the Consumer's behavior.

In the future, when OPS shows the asynchronous situation of the provider, we just need to let the provider still register the method.async property.

What do u think?

**Comment by diecui1202** [comment]
> After looking at the code, I think we really have to consider compatibility issues.

@carryxyh Could u describe the scenario detailedly?

**Comment by diecui1202** [comment]
So the decision is ignore method.async both on the provider and consumer side?

**Comment by carryxyh** [comment]
Yes, that is the best way for this bug.

**Comment by chickenlj** [comment]
> Therefore, for compatibility reasons, it is necessary to fix the problem on the Provider side so that method.async is not registered to the registry.

We are working on this in branch dev-meta.

> In the Consumer, ignoring the Provider's method.async is equivalent to an insurance, ensuring that the Provider's method.async will not affect the Consumer's behavior under any circumstances.

I have seen that you fixed this part in #2434. I tend to merge it now.
